### PR TITLE
loader: Support ABGR colorspace

### PR DESCRIPTION
**Pull request by JSUYA**
Since the color space is set at the time of specifying the target buffer of the canvas, there is no way to know the color space when the picture is loaded. So, check the color space applied to SwCanvas at the time of reload() and change the color space.

There is an issue of BGR color space support for each loader. The external_jpg loader resets the TJPF color space and calls read() to get a new buffer. In the case of external_png, we need to change the color value directly because it have to start over from begin_read_*.

This solution can affect performance as much as it access again image buffer that have already been `read()` done. However, this only happens once.

issue: https://github.com/Samsung/thorvg/issues/1310

### CODE REVIEW

**Comment by bchrescionko** [comment]
I checked it on svg test file and now it works fine, colors are appropriate for different color spaces.

File: src/lib/tvgPictureImpl.h [hunk 1]
```diff
@@ -89,7 +89,7 @@ struct Picture::Impl
         return ret;
     }
 
-    uint32_t reload()
+    uint32_t reload(int32_t colorSpace)
```
Comment (hermet) [inline src/lib/tvgPictureImpl.h:92]: Suggest setting the default value here.
uint32_t reload(uint32_t colorSpace = -1)

File: src/loaders/external_jpg/tvgJpgLoader.h [hunk 2]
```diff
@@ -48,6 +50,7 @@ class JpgLoader : public LoadModule
     unsigned char *image = nullptr;
     unsigned long size = 0;
     bool freeData = false;
+    uint32_t tjpfColorSpace = TJPF_BGRX;
```
Comment (hermet) [inline src/loaders/external_jpg/tvgJpgLoader.h:53]: I think a new variable is not necessary because Loader already has a colorSpace variable.
you can add a local converter function and immediately call it when it's necessary (see mentioned the above comment)

File: src/loaders/external_jpg/tvgJpgLoader.cpp [hunk 3]
```diff
@@ -149,16 +148,32 @@ bool JpgLoader::close()
 }
 
 
-unique_ptr<Surface> JpgLoader::bitmap()
+unique_ptr<Surface> JpgLoader::bitmap(int32_t colorSpace)
 {
     if (!image) return nullptr;
+    if (colorSpace != -1 && static_cast<uint32_t>(colorSpace) != this->colorSpace) {
```
Comment (hermet) [inline src/loaders/external_jpg/tvgJpgLoader.cpp:154]: probably you can make a separate converter that returns converted value. In this case, you can call it immediately when it's necessary, no need an additional member variable.
```
uint32_t convert(uint32_t colorSpace)
{
   uint32 tjpfColorSpace;
    ...
   return tjpfColorSpace;
}
```

File: src/loaders/external_png/tvgPngLoader.cpp [hunk 4]
```diff
@@ -110,16 +130,20 @@ bool PngLoader::close()
     return true;
 }
 
-unique_ptr<Surface> PngLoader::bitmap()
+unique_ptr<Surface> PngLoader::bitmap(int32_t colorSpace)
 {
     if (!content) return nullptr;
+    if (colorSpace != -1 && static_cast<uint32_t>(colorSpace) != this->colorSpace) {
```
Comment (hermet) [inline src/loaders/external_png/tvgPngLoader.cpp:136]: This converting compatibility looks quite fragile. At least you can add a description here(other loaders as well) what it does and how this must be taken over the possible coming other colorspace types.

File: src/loaders/external_png/tvgPngLoader.h [hunk 5]
```diff
@@ -37,11 +37,12 @@ class PngLoader : public LoadModule
...
     bool close() override;
 
-    unique_ptr<Surface> bitmap() override;
+    unique_ptr<Surface> bitmap(int32_t colorSpace) override;
 
 private:
     png_imagep image = nullptr;
-    const uint32_t* content = nullptr;
+    uint32_t* content = nullptr;
+    int pngFormat = PNG_FORMAT_BGRA;
```
Comment (hermet) [inline src/loaders/external_png/tvgPngLoader.h:45]: we can remove this as well??

File: src/loaders/external_png/tvgPngLoader.cpp [hunk 6]
```diff
@@ -97,6 +116,7 @@ bool PngLoader::read()
         free(buffer);
         return false;
     }
+    if (content) free((void*)content);
```
Comment (hermet) [inline src/loaders/external_png/tvgPngLoader.cpp:119]: this looks an unrelated change... does this correct?

File: src/loaders/external_jpg/tvgJpgLoader.cpp (hunk 3 above)
Comment (JSUYA) [reply src/loaders/external_jpg/tvgJpgLoader.cpp]: thank you for your advice. i added convert function.

File: src/loaders/external_png/tvgPngLoader.cpp (hunk 6 above)
Comment (JSUYA) [reply src/loaders/external_png/tvgPngLoader.cpp]: removed

File: src/lib/tvgRender.h [hunk 7]
```diff
@@ -37,7 +37,7 @@ struct Surface
     uint32_t* buffer;
     uint32_t  stride;
     uint32_t  w, h;
-    uint32_t  cs;
+    uint32_t  colorSpace;
```
Comment (hermet) [inline src/lib/tvgRender.h:40]: This is actually out of this goal. Next time please commit an individual refactoring change before the original mission.

**Comment by hermet** [comment]
Unnecessary commits are welcomed. I'd like to suggest you to use "git rebase --interactive" to update the commits in the existing PR. Otherwise all commits will be squashed at the end.

File: src/lib/tvgLoadModule.h [hunk 8]
```diff
@@ -49,7 +50,7 @@ class LoadModule
 
     virtual bool read() = 0;
     virtual bool close() = 0;
-    virtual unique_ptr<Surface> bitmap() { return nullptr; }
+    virtual unique_ptr<Surface> bitmap(bool changeColorSpace) { return nullptr; }
```
Comment (hermet) [inline src/lib/tvgLoadModule.h:53]: Thinking about this... 
changeColorSpace is too ambiguous. In the end, the requested colorspace value has to be specified. 

How about this?
```
enum Colorspace { RGBA, ... };
class Coverter { 
  static RGBA2ABGR();
  ...
 };  //in tvgRender.h
...
LoadModule::bitmap(Colorspace cs);
```

File: src/loaders/external_png/tvgPngLoader.cpp [hunk 9]
```diff
@@ -42,6 +42,24 @@ static void _premultiply(uint32_t* data, uint32_t w, uint32_t h)
 }
 
 
+static inline uint32_t CHANGE_COLORSPACE(uint32_t c)
```
Comment (hermet) [inline src/loaders/external_png/tvgPngLoader.cpp:45]: We don't need this duplicate logic if you consider the general Converter mentioned above.

**Review by hermet — CHANGES_REQUESTED** [review]
More feedbacks. please check it. thank you.

File: src/lib/tvgRender.h (hunk 7 above)
Comment (JSUYA) [reply src/lib/tvgRender.h]: reverted the code

File: src/lib/tvgLoadModule.h (hunk 8 above)
Comment (JSUYA) [reply src/lib/tvgLoadModule.h]: I changed the bool changeColorSpace to the colorSpace value and compare it inside the loader.
The colorspace change used now is ARGB <-> ABGR.
ARGBtoABGR or ABGRtoARGB have the same logic, so I think, The current logic doesn't have enough functionality to make it a class.
Some of the common functions can be moved to the header, but I think it is in the same context as _premultiply(), and after this PR, we can work on a separate PR to organize these.

Comment (hermet) [reply src/lib/tvgLoadModule.h]: Ok, actually both are spagetties. Back then I allowed it because I have no better idea at that time. Even now. (Do you?) However colorspace converter is different. changeColorSpace() with out any specification is too ambugiuous because there will be more colorspace formats such as RGB, ARGB, BGR, BGRA, GRAYSCALE, etc. 

It's my opinion. Overall, I agree on your contibution code, anyway it helps ThorVG better than before. But I would like to let you know there is a possibibility to revise the whole base code.

**Review by hermet — APPROVED** [review]
thanks.
